pkg/apis/provisioning/v0alpha1/types.go

```diff
@@ -0,0 +1,110 @@
+package v0alpha1
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Repository struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec RepositorySpec `json:"spec,omitempty"`
+}
+
+type LocalRepository struct {
+	Path string `json:"path,omitempty"`
+}
+
+func (r *LocalRepository) IsEmpty() bool {
+	return r == nil || r.Path == ""
+}
+
+type S3Repository struct {
+	Bucket string `json:"bucket,omitempty"`
+
+	// TODO: Add ACL?
+	// TODO: Encryption??
+	// TODO: How do we define access? Secrets?
+}
+
+func (r *S3Repository) IsEmpty() bool {
+	return r == nil || r.Bucket == ""
+}
+
+type GitHubRepository struct {
+	// The owner of the repository (e.g. example in `example/test` or `https://github.com/example/test`).
+	Owner string `json:"owner,omitempty"`
+	// The name of the repository (e.g. test in `example/test` or `https://github.com/example/test`).
+	Repository string `json:"repository,omitempty"`
+
+	// TODO: Do we want an SSH url instead maybe?
+	// TODO: On-prem GitHub Enterprise support?
+	// TODO: How do we define access? Secrets?
+
+	// Whether we should commit to change branches and use a Pull Request flow to achieve this.
+	// By default, this is false (i.e. we will commit straight to the main branch).
+	BranchWorkflow bool `json:"branchWorkflow,omitempty"`
+
+	// Whether we should show dashboard previews in the pull requests caused by the BranchWorkflow option.
+	// By default, this is false (i.e. we will not create previews).
+	// This option is a no-op if BranchWorkflow is `false` or default.
+	GenerateDashboardPreviews bool `json:"generateDashboardPreviews,omitempty"`
+}
+
+func (r *GitHubRepository) IsEmpty() bool {
+	// we don't need to check options here, just the most important stuff to actually connect.
+	return r == nil || r.Owner == "" || r.Repository == ""
+}
+
+type RepositorySpec struct {
+	// The UID of the folder that is backed by the repository.
+	FolderUID string `json:"folderUid,omitempty"`
```

Comment:
```suggestion
	Folder string `json:"folder,omitempty"`
```
Lets avoid "UID" if possible -- it has too many confusing overlaps.  In k8s, this will be the "name" for the folder we want to write to (and UID is not something stable across deployments)